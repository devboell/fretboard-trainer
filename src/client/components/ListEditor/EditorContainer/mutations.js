import gql from 'graphql-tag'

export const CREATE_QUIZ = gql`
  mutation CreateQuiz(
    $name: String,
    $type: QuizTypeEnum,
    $tuning: TuningEnum,
    $width: Int
  ) {
    createQuiz(
      name: $name,
      type: $type,
      tuning: $tuning,
      width: $width
    ) {
      id
      name
      type
      tuning
      width
    }
  }
`

export const UPDATE_QUIZ = gql`
  mutation UpdateQuiz(
    $id: ID,
    $name: String,
    $type: QuizTypeEnum,
    $tuning: TuningEnum,
    $width: Int
  ) {
    updateQuiz(
      id: $id,
      name: $name,
      type: $type,
      tuning: $tuning,
      width: $width
    ) {
      id
      name
      type
      tuning
      width
    }
  }
`

export const DELETE_QUIZ = gql`
  mutation DeleteQuiz($id: ID) {
    deleteQuiz(id: $id)
  }
`
