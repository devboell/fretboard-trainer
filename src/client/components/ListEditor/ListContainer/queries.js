import gql from 'graphql-tag'

export default gql`
  {
    quizzes {
      id
      name
      type
      tuning
      width
    }
  }
`

