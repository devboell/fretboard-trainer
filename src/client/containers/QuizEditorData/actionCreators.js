export const setSelectedQuizId = id => ({
  type: 'SET_SELECTED_QUIZ_ID',
  id,
})

export const setIsNew = flag => ({
  type: 'SET_IS_NEW',
  flag,
})

export const setEditedQuiz = quiz => ({
  type: 'SET_EDITED_QUIZ',
  quiz,
})

export const updateEditedQuiz = (key, value) => ({
  type: 'UPDATE_EDITED_QUIZ',
  key,
  value,
})
