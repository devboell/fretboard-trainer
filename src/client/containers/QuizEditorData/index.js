import React from 'react'
import pt from 'prop-types'
import { ApolloConsumer } from 'react-apollo'

import QuizEditorPage from 'components/QuizEditorPage'
import withData from './enhancers'

const QuizEditorData = ({
  loading,
  error,
  ...other
}) => {
  if (loading) return <p>Loading...</p>
  if (error) return <p>Error</p>

  return (
    <ApolloConsumer>
      {(cache) => {
        const selectQuiz = id => cache.writeData({ data: { isNew: false, selectedQuizId: id } })
        const clearQuizSelection = isNewFlag => cache.writeData({ data: { isNew: isNewFlag, selectedQuizId: 'no_selection' } })

        return (
          <QuizEditorPage
            {...{ ...other, selectQuiz, clearQuizSelection }}
          />
      )
    }}
    </ApolloConsumer>
  )
}

QuizEditorData.propTypes = {
  loading: pt.bool.isRequired,
  error: pt.shape({}),
}

QuizEditorData.defaultProps = {
  error: undefined,
}

export default withData(QuizEditorData)
