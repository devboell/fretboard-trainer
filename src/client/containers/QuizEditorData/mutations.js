import gql from 'graphql-tag'

export const CREATE_QUIZ = gql`
  mutation CreateQuiz($name: String) {
    createQuiz(name: $name) {
      id
      name
    }
  }
`

export const UPDATE_QUIZ = gql`
  mutation UpdateQuiz($id: ID, $name: String) {
    updateQuiz(id: $id, name: $name) {
      id
      name
    }
  }
`

export const DELETE_QUIZ = gql`
  mutation DeleteQuiz($id: ID) {
    deleteQuiz(id: $id)
  }
`
