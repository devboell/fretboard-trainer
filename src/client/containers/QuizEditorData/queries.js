import gql from 'graphql-tag'

export const QUIZZES = gql`
  {
    quizzes {
      id
      name
      type
    }
  }
`
export const CLIENT_UI = gql`
  {
    selectedQuizId @client
    isNew @client
  }
`
