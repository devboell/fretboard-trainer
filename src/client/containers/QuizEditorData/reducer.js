import { lensProp, set } from 'ramda'

const initialState = {
  selectedQuizId: 'no_selection',
  isNew: false,
  editedQuiz: undefined,
}

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_SELECTED_QUIZ_ID':
      return { ...state, selectedQuizId: action.id }
    case 'SET_IS_NEW':
      return { ...state, isNew: action.flag }
    case 'SET_EDITED_QUIZ':
      return { ...state, editedQuiz: action.quiz }
    case 'UPDATE_EDITED_QUIZ': {
      const { key, value } = action
      const lensKey = lensProp(key)
      const { editedQuiz } = state
      return {
        ...state,
        editedQuiz: set(lensKey, value, editedQuiz),
      }
    }

    default:
      return state
  }
}
