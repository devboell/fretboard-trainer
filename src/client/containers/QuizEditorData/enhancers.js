import { graphql, compose } from 'react-apollo'

import { QUIZZES, CLIENT_UI } from './queries'
import { CREATE_QUIZ, UPDATE_QUIZ, DELETE_QUIZ } from './mutations'


const quizzesQueryProps = ({ data: { quizzes, loading, error } }) => ({
  quizzes, loading, error,
})

const withQuizzesQuery = graphql(
  QUIZZES,
  { props: quizzesQueryProps },
)

const clientUIProps = ({ data: { selectedQuizId, isNew } }) => ({
  selectedQuizId, isNew,
})

const withClientUIQuery = graphql(
  CLIENT_UI,
  { props: clientUIProps },
)

const createQuizProps = ({ mutate }) => ({
  createQuiz: values => mutate({
    variables: values,
  }),
})

const createQuizQueryUpdate = (proxy, { data: { createQuiz } }) => {
  const { quizzes } = proxy.readQuery({ query: QUIZZES })
  const result = [...quizzes, createQuiz]

  proxy.writeQuery({ query: QUIZZES, data: { quizzes: result } })
  proxy.writeQuery({
    query: CLIENT_UI,
    data: {
      selectedQuizId: createQuiz.id,
      isNew: false,
    },
  })
}

const withCreateQuizMutation = graphql(
  CREATE_QUIZ,
  {
    props: createQuizProps,
    options: {
      update: createQuizQueryUpdate,
    },
  },
)

const updateQuizProps = ({ mutate }) => ({
  updateQuiz: values => mutate({
    variables: values,
  }),
})

const withUpdateQuizMutation = graphql(
  UPDATE_QUIZ,
  { props: updateQuizProps },
)

const deleteQuizProps = ({ mutate }) => ({
  deleteQuiz: values => mutate({
    variables: values,
  }),
})

const deleteQuizQueryUpdate = (proxy, { data: { deleteQuiz } }) => {
  const { quizzes } = proxy.readQuery({ query: QUIZZES })
  const result = quizzes.filter(q => q.id !== deleteQuiz)

  proxy.writeQuery({ query: QUIZZES, data: { quizzes: result } })
}

const withDeleteQuizMutation = graphql(
  DELETE_QUIZ,
  {
    props: deleteQuizProps,
    options: {
      update: deleteQuizQueryUpdate,
    },
  },
)


export default compose(
  withQuizzesQuery,
  withClientUIQuery,
  withUpdateQuizMutation,
  withCreateQuizMutation,
  withDeleteQuizMutation,
)
