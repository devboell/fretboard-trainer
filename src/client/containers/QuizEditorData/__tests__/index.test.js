import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import { contains, find, propEq, eqProps, equals } from 'ramda'
import testData from 'fixtures/quizEditor'
import QuizEditorData from '../index'
import { QUIZZES, CLIENT_UI } from '../queries'
import { CREATE_QUIZ, UPDATE_QUIZ, DELETE_QUIZ } from '../mutations'

const createResult = { id: '7', name: 'new' }
const updateResult = { id: '1', name: 'updated', __typename: 'Quiz' }
const deleteResult = '1'

const mutationMocks = [
  {
    request: { query: CREATE_QUIZ },
    result: { data: { createQuiz: createResult } },
  },
  {
    request: { query: UPDATE_QUIZ },
    result: { data: { updateQuiz: updateResult } },
  },
  {
    request: { query: DELETE_QUIZ },
    result: { data: { deleteQuiz: deleteResult } },
  },
]

const updateChildProps = wrapper =>
  wrapper.update().find('QuizEditorPage').props()

describe('QuizEditorData with default state', () => {
  let wrapper
  let childProps
  const data = testData('no_selection', false)

  const queryMocks = [
    {
      request: { query: QUIZZES },
      result: { data: { quizzes: data.quizzes } },
    },
    {
      request: { query: CLIENT_UI },
      result: {
        data: {
          selectedQuizId: data.selectedQuizId,
          isNew: data.isNew,
        },
      },
    },
  ]

  beforeEach(async () => {
    wrapper = mount((
      <MockedProvider
        removeTypename
        mocks={[...queryMocks, ...mutationMocks]}
      >
        <QuizEditorData />
      </MockedProvider>))
    // makes sure the loading is done
    await new Promise(resolve => setTimeout(resolve))
    childProps = updateChildProps(wrapper)
  })


  it('gives the child component initial loaded props', () => {
    expect(childProps).toMatchSnapshot()
  })

  it('calls createQuiz and updates the cache correctly', async () => {
    await childProps.createQuiz()
    childProps = updateChildProps(wrapper)
    const result = contains(createResult, childProps.quizzes)
      && equals(createResult.id, childProps.selectedQuizId)

    expect(result).toBeTruthy()
  })

  it('calls updateQuiz and updates the cache correctly', async () => {
    await childProps.updateQuiz()
    childProps = updateChildProps(wrapper)
    const updatedQuiz = find(propEq('id', updateResult.id))(childProps.quizzes)
    const result = eqProps('name', updateResult, updatedQuiz)

    expect(result).toBeTruthy()
  })

  it('calls deleteQuiz and updates the cache correctly', async () => {
    await childProps.deleteQuiz()
    childProps = updateChildProps(wrapper)
    const result = find(propEq('id', deleteResult))(childProps.quizzes)
    expect(result).toBeUndefined()
  })

  it('calls selectQuiz and updates the cache correctly', async () => {
    const selectedId = '3'
    await childProps.selectQuiz(selectedId)
    childProps = updateChildProps(wrapper)
    const result = equals(childProps.selectedQuizId, selectedId)

    expect(result).toBeTruthy()
  })

  it('calls clearQuizSelection(true) and updates the cache correctly', async () => {
    const flag = true
    await childProps.clearQuizSelection(flag)
    childProps = updateChildProps(wrapper)
    const result = equals(childProps.isNew, flag)

    expect(result).toBeTruthy()
  })
})

describe('QuizEditorData loading and error state', () => {
  it('shows loading element', () => {
    const wrapper = mount((
      <MockedProvider
        removeTypename
      >
        <QuizEditorData />
      </MockedProvider>))
    expect(wrapper.find('QuizEditorData')).toMatchSnapshot()
  })

  it('shows error element', async () => {
    const wrapper = mount((
      <MockedProvider
        removeTypename
        mocks={{
          request: { query: QUIZZES },
          result: { data: { error: 'err msg' } },
        }}
      >
        <QuizEditorData />
      </MockedProvider>))

    await new Promise(resolve => setTimeout(resolve))
    expect(wrapper.update().find('QuizEditorData')).toMatchSnapshot()
  })
})

