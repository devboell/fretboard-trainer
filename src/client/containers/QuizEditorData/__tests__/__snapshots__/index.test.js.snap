// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizEditorData loading and error state shows error element 1`] = `
<QuizEditorData
  createQuiz={[Function]}
  deleteQuiz={[Function]}
  error={
    [Error: Network error: No more mocked responses for the query: {
  quizzes {
    id
    name
    type
  }
}
, variables: {}]
  }
  loading={false}
  updateQuiz={[Function]}
>
  <p>
    Error
  </p>
</QuizEditorData>
`;

exports[`QuizEditorData loading and error state shows loading element 1`] = `
<QuizEditorData
  createQuiz={[Function]}
  deleteQuiz={[Function]}
  loading={true}
  updateQuiz={[Function]}
>
  <p>
    Loading...
  </p>
</QuizEditorData>
`;

exports[`QuizEditorData with default state gives the child component initial loaded props 1`] = `
Object {
  "clearQuizSelection": [Function],
  "createQuiz": [Function],
  "deleteQuiz": [Function],
  "isNew": false,
  "quizzes": Array [
    Object {
      "id": "1",
      "name": "quiz one",
      "type": "pc",
      Symbol(id): "Quiz:1",
    },
    Object {
      "id": "2",
      "name": "quiz two",
      "type": "pc",
      Symbol(id): "Quiz:2",
    },
    Object {
      "id": "3",
      "name": "quiz three",
      "type": "pc",
      Symbol(id): "Quiz:3",
    },
    Object {
      "id": "4",
      "name": "quiz four",
      "type": "pc",
      Symbol(id): "Quiz:4",
    },
    Object {
      "id": "5",
      "name": "quiz five",
      "type": "pc",
      Symbol(id): "Quiz:5",
    },
    Object {
      "id": "6",
      "name": "quiz six",
      "type": "pc",
      Symbol(id): "Quiz:6",
    },
  ],
  "selectQuiz": [Function],
  "selectedQuizId": "no_selection",
  "updateQuiz": [Function],
}
`;
